docs/build/guides/conventions/analyzing-smart-contract-cost.mdx

```diff
@@ -0,0 +1,449 @@
+    }
+}
+
+```
+
+:::tip
+
+In this optimized approach, we'll accumulate the changes outside the loop and perform a single storage update and also a single read operation. This reduces the number of storage operations and the overall gas cost of the contract.
+
+:::
+
+## 2. **Use of Events over Storage**
+
+Events are a cost-effective way to store data that doesn't need to be accessed frequently. Events are cheaper than storage operations and can be used to store data that doesn't need to be accessed frequently.
+
+#### Not Optimal Contract ❎
+
+```rust
+#![no_std]
+use soroban_sdk::{
+    contract, contractimpl, log, symbol_short, vec, Address, Env, Symbol,Map, Vec
+};
+
+#[contract]
+pub struct GameContract;
+
+#[contractimpl]
+impl GameContract {
+    // Function to record a game move and update storage
+    pub fn record_move(env: Env, player: Address, move_type: Symbol) {
+        let mut player_moves: Map<Address, Vec<Symbol>> = env.storage().persistent().get("player_moves");
+        let moves = player_moves.get(&player);
+        moves.push(move_type.clone());
+        player_moves.set(player, moves);
+        env.storage().persistent().set("player_moves", &player_moves);
+    }
+
+    // Function to unlock an achievement and update storage
+    pub fn unlock_achievement(env: Env, player: Address, achievement: Symbol) {
+        let mut player_achievements: Map<Address, Vec<Symbol>> = env.storage().persistent().get("player_achievements");
+        let achievements = player_achievements.get(&player);
+        achievements.push(achievement.clone());
+        player_achievements.set(player, achievements);
+        env.storage().persistent().set("player_achievements", &player_achievements);
+    }
+
+}
+```
+
+:::danger
+
+We cannot store everything in storage like we would do in a traditional database. This approach is not cost-effective as it involves multiple storage operations for each player move and saves each achievement to the storage
+
+:::
+
+#### Optimal Contract ✅
+
+```rust
+#![no_std]
+use soroban_sdk::{
+    contract, contractimpl, log, symbol_short, vec, Address, Env, Symbol,Map, Vec,
+};
+
+#[contract]
+pub struct GameContract;
+
+#[contractimpl]
+impl GameContract {
+    // Function to record a game move and emit an event
+    pub fn record_move(env: Env, player: Address, move_type: Symbol) {
+        // Emit event for the game move
+        env.events().publish(("game_move",), (&player, move_type.clone()));
+    }
+
+    // Function to unlock an achievement and emit an event
+    pub fn unlock_achievement(env: Env, player: Address, achievement: Symbol) {
+        // Emit event for the unlocked achievement
+        env.events().publish(("achievement_unlocked",), (&player, achievement.clone()));
+    }
+}
+```
+
+:::tip
+
+In this optimized approach, we'll use events to store data that doesn't need to be accessed frequently. This reduces the number of storage operations and the overall gas cost of the contract.
```

Comment:
I'd like to see a discussion of trade-offs here, as well. For example, you could store the latest move in persistent storage and the rest of history in temporary storage if you didn't want ephemeral events to be the only option. You wouldn't have to describe the code for that, but a quick discussion here would be really useful.

Reply:
In fact the approach of storing `Map<Address, N>` is not only inefficient, it might even be impossible because of the "caps" on the size of a single storage slot. With that in mind, another alternative I'd like to see here, if someone where to want to stick to the storage-based approach, would be to use the `Address` as the storage key itself and the moves/achievements as the result:

```rust
pub enum DataKey { 
  Moves(Address),
  Achievements(Address)
}

// later...

env.storage().persistent().get(DataKey::Moves(player), &movelist);
```

and so forth.

Reply:
So maybe you could have different "Optimal Contract" sections for this outlining the spectrum of ways you could tackle the inefficient storage problem.